tsdb/exemplar_test.go

```diff
@@ -119,11 +120,521 @@ func TestAddExemplar(t *testing.T) {
+				{Labels: series1, Value: 0.2, Ts: 2},
+			},
+			wantNextIndex: 0,
+		},
+		{
+			name: "shrink to two",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 1},
+			},
+			resize: 2,
+			wantExemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 1},
+			},
+			wantNextIndex: 1,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			exs, err := NewCircularExemplarStorage(3, eMetrics, 100)
+			require.NoError(t, err)
+			es := exs.(*CircularExemplarStorage)
+
+			for _, ex := range tc.exemplars {
+				require.NoError(t, es.AddExemplar(ex.Labels, ex))
+			}
+
+			// Resize the circular buffer.
+			if testing.Verbose() {
+				t.Logf("Buffer[before-resize]:\n%s", debugCircularBuffer(es))
+			}
+			es.Resize(tc.resize)
+			if testing.Verbose() {
+				t.Logf("Buffer[after-resize]:\n%s", debugCircularBuffer(es))
+			}
+
+			// Ensure exemplars are returned correctly and in-order.
+			gotExemplars, err := es.Select(0, 1000, matcher1)
+			require.NoError(t, err)
+			flat := make([]exemplar.Exemplar, 0)
+			for _, group := range gotExemplars {
+				flat = append(flat, group.Exemplars...)
+			}
+			sort.Slice(flat, func(i, j int) bool {
+				return flat[i].Ts < flat[j].Ts
+			})
+			require.Equal(t, tc.wantExemplars, flat, "exemplar mismatch")
+			require.Equal(t, tc.wantNextIndex, es.nextIndex, "next index mismatch")
+		})
+	}
+
+	exemplars1to4 := []exemplar.Exemplar{
+		{Labels: series1, Value: 0.1, Ts: 1},
+		{Labels: series1, Value: 0.2, Ts: 2},
+		{Labels: series1, Value: 0.3, Ts: 3},
+		{Labels: series1, Value: 0.4, Ts: 4},
+	}
+	exemplars5to7 := []exemplar.Exemplar{
+		{Labels: series1, Value: 0.5, Ts: 5},
+		{Labels: series1, Value: 0.6, Ts: 6},
+		{Labels: series1, Value: 0.7, Ts: 7},
+	}
+
+	t.Run("shrink then grow", func(t *testing.T) {
+		exs, err := NewCircularExemplarStorage(3, eMetrics, 100)
+		require.NoError(t, err)
+		es := exs.(*CircularExemplarStorage)
+
+		for _, ex := range exemplars1to4 {
+			require.NoError(t, es.AddExemplar(ex.Labels, ex))
+		}
+
+		if testing.Verbose() {
+			t.Logf("Buffer[before-resize-2]:\n%s", debugCircularBuffer(es))
+		}
+
+		es.Resize(2)
+		if testing.Verbose() {
+			t.Logf("Buffer[after-resize-2]:\n%s", debugCircularBuffer(es))
+		}
+
+		gotExemplars, err := es.Select(0, 1000, matcher1)
+		require.NoError(t, err)
+		require.Len(t, gotExemplars, 1)
+		require.Len(t, gotExemplars[0].Exemplars, 2)
+
+		if testing.Verbose() {
+			t.Logf("Buffer[before-resize-5]:\n%s", debugCircularBuffer(es))
+		}
+		es.Resize(5)
+
+		if testing.Verbose() {
+			t.Logf("Buffer[after-resize-5]:\n%s", debugCircularBuffer(es))
+		}
+		for _, ex := range exemplars5to7 {
+			require.NoError(t, es.AddExemplar(ex.Labels, ex))
+		}
+		gotExemplars, err = es.Select(0, 1000, matcher1)
+		require.NoError(t, err)
+		require.Len(t, gotExemplars, 1)
+		require.Len(t, gotExemplars[0].Exemplars, 5)
+	})
+
+	t.Run("grow then shrink", func(t *testing.T) {
```

Comment:
can you add a two more test cases - shrink & grow + out of order and the grow & shrink + out of order?

idk if it wouldn't be easier to just extend the test cases above - you'd have something like 


```go
testCases := []struct {
    name          string
    exemplars     []exemplar.Exemplar
    wantExemplarsBeforeResize []exemplar.Exemplar
    resize        int64
    exemplarsAfterResize []exemplar.Exemplar
    wantExemplarsAfterResize []exemplar.Exemplar
    wantError     error
}
```